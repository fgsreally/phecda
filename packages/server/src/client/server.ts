import type amqplib from 'amqplib'
export * from './axios'

export function createMqReq(channel: amqplib.Channel): <R>(arg: R) => Promise<void> {
  return async (arg: any) => {
    const { url, body } = arg
    await channel.assertQueue(url)
    await channel.sendToQueue(url, Buffer.from(JSON.stringify(body)))
  }
}
