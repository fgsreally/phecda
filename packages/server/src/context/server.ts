import type { ValidatePipe } from '../pipe'
import { defaultPipe } from '../pipe'
import type { ErrorFilter } from '../filter'
import { serverFilter } from '../filter'
import { WrongMetaException } from '../exception/wrong-meta'
import { Pcontext } from './base'

export class ServerContext extends Pcontext {
  static pipe = defaultPipe
  static filter = serverFilter
  static middlewareRecord: Record<string, (...params: any) => any> = {}
  static useMiddleware(middlewares: string[]) {
    return middlewares.map((m) => {
      if (!(m in ServerContext.middlewareRecord))
        throw new WrongMetaException(`can't find middleware named ${m}`)
      return ServerContext.middlewareRecord[m]
    })
  }

  async usePipe(args: { arg: any; validate?: boolean }[], reflect: any[]) {
    return ServerContext.pipe.transform?.(args, reflect)
  }

  useFilter(arg: any) {
    return ServerContext.filter(arg, this.data)
  }
}

export function addMiddleware(key: string, handler: (...params: any) => any) {
  ServerContext.middlewareRecord[key] = handler
}

export function useServerPipe(pipe: ValidatePipe) {
  ServerContext.pipe = pipe
}
export function useServerFilter(filter: ErrorFilter) {
  ServerContext.filter = filter
}
