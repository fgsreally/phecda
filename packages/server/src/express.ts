import type { Express } from 'express'
import { PhecdaServer } from './server'
import type { ServerMeta } from './types'
import { HttpException } from './exception/base'
import { isObject } from './utils'
export function bindApp(app: Express, { meta, moduleMap }: { meta: ServerMeta[]; moduleMap: any }, key = '/__PHECDA_SERVER__') {
  const methodMap = {} as Record<string, (...args: any[]) => any>
  for (const i of meta) {
    const server = new PhecdaServer(`${i.name}`, i)
    const instance = moduleMap.get(i.name)!
    const method = server.requestToMethod(instance[i.method].bind(instance))
    methodMap[`${i.name}-${i.method}`] = method
    if (i.route.type) {
      app[i.route.type](i.route.route, async (req, res) => {
        const ret = await method(req)
        if (ret instanceof HttpException) {
          res.status(ret.status)
          res.json(ret)
        }
        if (isObject(ret))
          res.json(ret)
        else
          res.send(String(ret))
      })
    }
  }
  app.post(key, async (req, res) => {
    const { body } = req
    const ret = [] as any[]
    for (const i in body)
      ret.push(await methodMap[i](body[i]))

    res.json(ret)
  })
}
