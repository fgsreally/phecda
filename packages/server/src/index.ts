import 'reflect-metadata'

type Construct<T = any> = new (...args: Array<any>) => T

export function Factory<T>(Module: Construct<T>) {
  const moduleMap = new Map<Construct, InstanceType<Construct>>()
  const app = buildNestModule(Module, moduleMap) as InstanceType<Construct<T>>
  return { app, moduleMap }
}

function buildNestModule(Module: Construct, map: Map<Construct, InstanceType<Construct>>) {
  const paramtypes = getParamtypes(Module) as Construct[]
  let instance: InstanceType<Construct>
  if (map.has(Module)) {
    instance = map.get(Module)
    if (!instance)
      throw new Error(`exist Circular Module dep--${Module}`)
    return instance
  }
  map.set(Module, undefined)
  if (paramtypes) {
    instance = new Module(...paramtypes.map(item =>
      buildNestModule(item, map),
    ))
  }
  else {
    instance = new Module()
  }
  map.set(Module, instance)

  return instance
}

function getParamtypes(Module: Construct) {
  return Reflect.getMetadata('design:paramtypes', Module)
}
