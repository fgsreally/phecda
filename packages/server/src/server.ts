import { HttpException } from './exception/base'
import { UndefinedException } from './exception/undefine'
import type { Meta } from './meta'
import type { PhecdaPipe } from './pipe'
import { defaultPipe } from './pipe'

export class PhecdaServer {
  method: string
  params: string[]
  static pipe: PhecdaPipe = defaultPipe
  static guardsRecord: Record<string, (...params: any) => boolean> = {}
  static interceptorsRecord: Record<string, (...params: any) => any | ((...params: any) => any)> = {}
  static serverRecord: Record<string, PhecdaServer> = {}

  constructor(public key: string, public meta: Meta) {
    PhecdaServer.serverRecord[key] = this
  }

  static registerGuard(key: string, handler: any) {
    PhecdaServer.guardsRecord[key] = handler
  }

  static registerInterceptor(key: string, handler: any) {
    PhecdaServer.interceptorsRecord[key] = handler
  }

  async useGuard(req: any, guards: string[]) {
    for (const guard of guards) {
      if (!await PhecdaServer.guardsRecord[guard]?.(req))
        throw new Error('aa')
    }
  }

  async useInterceptor(req: any, interceptors: string[]) {
    const ret = []
    for (const interceptor of interceptors) {
      const post = await PhecdaServer.interceptorsRecord[interceptor]?.(req)
      if (post)
        ret.push(post)
    }
    return ret
  }

  async usePost(ret: any, cbs: ((...params: any[]) => any)[]) {
    for (const cb of cbs)
      ret = (await cb(ret)) | ret

    return ret
  }

  async usePipe(args: { arg: any; validate: boolean }[], reflect: any[]) {
    return PhecdaServer.pipe.transform(args, reflect)
  }

  methodToHandler(method: (...params: any[]) => any) {
    const { data: { params = [], guards = [], interceptors = [] }, reflect = [] } = this.meta

    return async (req: any) => {
      try {
        await this.useGuard(req, guards)
        const posts = await this.useInterceptor(req, interceptors!)
        const args = params.map((param) => {
          return { arg: req[param.type]?.[param.key], validate: param.validate }
        })

        const ret = await method(...await this.usePipe(args, reflect))
        return this.usePost(ret, posts)
      }
      catch (e: any) {
        if (!(e instanceof HttpException))
          return new UndefinedException(e.message || e)
        return e
      }
    }
  }
}

export function addGuard(key: string, handler: (...params: any) => boolean) {
  PhecdaServer.registerGuard(key, handler)
}

export function addInterceptor(key: string, handler: (...params: any) => any | ((...params: any) => any)) {
  PhecdaServer.registerInterceptor(key, handler)
}

export function usePipe(pipe: PhecdaPipe) {
  PhecdaServer.pipe = pipe
}
