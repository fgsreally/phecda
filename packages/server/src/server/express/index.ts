export * from './context'
export * from './bind'
export * from './modules'
