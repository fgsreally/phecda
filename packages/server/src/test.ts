import type { Server } from 'node:http'
import type { SuperAgentTest, Test } from 'supertest'
import { type Construct, getTag } from 'phecda-core'
import { Factory } from './core'

import type { PickFunc } from './types'
import type { ControllerMeta } from './meta'
export async function TestFactory<T extends Construct[]>(...Modules: T) {
  const { moduleMap, modelMap } = await Factory(Modules)

  return {
    get<C extends T[number]>(Model: C): InstanceType<C> {
      const tag = getTag(Model)
      const module = moduleMap.get(tag)

      if (!module)
        throw new Error(`module "${String(tag)}" doesn't exist`)

      if (modelMap.get(module) !== Model)
        throw new Error(`module "${Model.name}" and "${String(tag)}" in modulemap are different modules`)

      return module
    },
  }
}

export type SuperTestRequest<T> = {
  [K in keyof T]: T[K] extends (...args: infer R) => any ? (...args: R) => Test : never;
}

export async function TestHttp(app: Server | any, { moduleMap, meta }: Awaited<ReturnType<typeof Factory>>, isAgent = true) {
  const { default: request, agent } = await import('supertest')
  const Agent = agent(app) as SuperAgentTest & { module: typeof module }

  function module<T extends Construct>(Module: T): SuperTestRequest<PickFunc<InstanceType<T>>> {
    const tag = getTag(Module)

    const module = moduleMap.get(tag)
    if (!module)
      throw new Error(`module "${String(tag)}" doesn't exist`)

    return new Proxy({}, {
      get(_target, p) {
        const { data } = (meta as ControllerMeta[]).find(({ data }) => data.name === Module.name && data.func === p && data.tag === tag)!
        return (...args: any) => {
          const ret = { body: {}, headers: {}, query: {}, func: data.http!.type, url: data.http!.prefix + data.http!.route } as any

          data.params.forEach((item) => {
            if (item.type === 'params') {
              ret.url = ret.url.replace(`:${item.key}`, args[item.index])
              return
            }
            if (item.type === 'query') {
              ret.query[item.key] = args[item.index]
              return
            }

            // body
            if (item.key)
              ret[item.type][item.key] = args[item.index]
            else
              ret[item.type] = args[item.index]
          })
          // @ts-expect-error @todo miss type
          let agent = (isAgent ? Agent : request(app))[ret.func](ret.url)
          if (Object.keys(ret.query).length > 0)
            agent = agent.query(ret.query)

          if (Object.keys(ret.headers).length > 0)
            agent = agent.set(ret.headers)

          if (Object.keys(ret.body).length > 0)
            agent = agent.send(ret.body)

          return agent
        }
      },
    }) as any
  }

  Agent.module = module

  return Agent
}
