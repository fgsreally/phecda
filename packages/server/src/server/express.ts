import type { Express } from 'express'
import { Pcontext, ServerContext, parseMeta } from '../context'
import { isObject, resolveDep } from '../utils'
import { MERGE_SYMBOL, SERIES_SYMBOL } from '../common'
import type { Factory } from '../core'
import { BadRequestException } from '../exception'
import type { Pmeta } from '../meta'
import type { ServerMergeCtx } from '../types'

export interface Options {
/**
 * 专用路由的值，默认为/__PHECDA_SERVER__，处理phecda-client发出的合并请求
 */
  route?: string
  /**
 * 全局守卫
 */
  globalGuards?: string[]
  /**
 * 全局拦截器
 */
  globalInterceptors?: string[]
  /**
 * 专用路由的中间件，全局中间件请在bindApp以外设置
 */
  middlewares?: string[]
}

export function bindApp(app: Express, { meta, moduleMap }: Awaited<ReturnType<typeof Factory>>, options: Options = {}) {
  const { globalGuards, globalInterceptors, route, middlewares: proMiddle } = { route: '/__PHECDA_SERVER__', globalGuards: [], globalInterceptors: [], middlewares: [], ...options } as Required<Options>
  const methodMap = {} as Record<string, (...args: any[]) => any>
  const contextMeta = {} as Record<string, Pmeta>
  for (const i of meta) {
    const { name, method, route, header, tag } = i.data
    const instance = moduleMap.get(tag)!
    const methodTag = `${tag}-${method}`
    contextMeta[methodTag] = i
    Pcontext.metaRecord[methodTag] = i
    let {
      guards,
      reflect,
      interceptors,
      params,
      middlewares,
    } = Pcontext.metaDataRecord[methodTag] ? Pcontext.metaDataRecord[methodTag] : (Pcontext.metaDataRecord[methodTag] = parseMeta(i))

    guards = [...globalGuards!, ...guards]
    interceptors = [...globalInterceptors!, ...interceptors]

    const handler = instance[method].bind(instance)
    methodMap[methodTag] = handler
    Pcontext.instanceRecord[name] = instance
    if (route) {
      app[route.type](route.route, ...ServerContext.useMiddleware(middlewares), async (req, res) => {
        const contextData = {
          request: req,
          meta: i,
          response: res,
        }
        const context = new ServerContext(methodTag, contextData)

        try {
          for (const name in header)
            res.set(name, header[name])
          await context.useGuard(guards)
          await context.useInterceptor(interceptors)
          const args = await context.usePipe(params.map(({ type, key, validate }) => {
            return { arg: resolveDep((req as any)[type], key), validate }
          }), reflect)
          instance.context = contextData

          const ret = await context.usePost(await handler(...args))
          if (isObject(ret))
            res.json(ret)
          else
            res.send(String(ret))
        }
        catch (e: any) {
          i.handlers.forEach(handler => handler.error?.(e))
          const err = await context.useFilter(e)
          res.status(err.status).json(err)
        }
      })
    }
  }

  app.post(route, (req, _res, next) => {
    (req as any)[MERGE_SYMBOL] = true
    next()
  }, ...ServerContext.useMiddleware(proMiddle), async (req, res) => {
    const { body: { category, data } } = req

    const contextData = {
      request: req,
      response: res,
      meta: contextMeta,
    } as unknown as ServerMergeCtx

    if (!Array.isArray(data))
      return res.json(await ServerContext.useFilter(new BadRequestException('data format should be an array'), contextData))

    contextData.tags = data.map((item: any) => item.tag)

    const context = new ServerContext(route, contextData)
    const ret = [] as any[]

    if (category === 'series') {
      for (const item of data) {
        const { tag } = item
        const [name] = tag.split('-')
        const {
          guards,
          reflect,
          interceptors,
          params,
        } = Pcontext.metaDataRecord[tag]
        const instance = moduleMap.get(name)

        try {
          if (!params)
            throw new BadRequestException(`"${tag}" doesn't exist`)

          await context.useGuard(guards, true)
          await context.useInterceptor(interceptors, true)
          const args = await context.usePipe(params.map(({ type, key, validate }) => {
            const arg = resolveDep(item[type], key)
            if (typeof arg === 'string' && arg.startsWith(SERIES_SYMBOL)) {
              const [, index, argKey] = arg.split('@')
              return { arg: resolveDep(ret[Number(index)], argKey || key), validate }
            }

            return { arg, validate }
          }), reflect) as any
          instance.context = contextData

          ret.push(await context.usePost(await methodMap[tag](...args)))
        }
        catch (e: any) {
          const m = Pcontext.metaRecord[tag]
          m.handlers.forEach(handler => handler.error?.(e))
          ret.push(await context.useFilter(e))
        }
      }
      return res.json(ret)
    }
    if (category === 'parallel') {
      return Promise.all(data.map((item: any) => {
        // eslint-disable-next-line no-async-promise-executor
        return new Promise(async (resolve) => {
          const { tag } = item
          const [name] = tag.split('-')
          const {
            guards,
            reflect,
            interceptors,
            params,
          } = Pcontext.metaDataRecord[tag]
          const instance = moduleMap.get(name)

          try {
            if (!params)
              throw new BadRequestException(`"${tag}" doesn't exist`)

            await context.useGuard(guards, true)
            await context.useInterceptor(interceptors, true)
            const args = await context.usePipe(params.map(({ type, key, validate }) => {
              const arg = resolveDep(item[type], key)
              return { arg, validate }
            }), reflect) as any
            instance.context = contextData
            resolve(await context.usePost(await methodMap[tag](...args)))
          }
          catch (e: any) {
            const m = Pcontext.metaRecord[tag]
            m.handlers.forEach(handler => handler.error?.(e))
            resolve(await context.useFilter(e))
          }
        })
      })).then((ret) => {
        res.json(ret)
      })
    }

    res.json(await context.useFilter(new BadRequestException('category should be \'parallel\' or \'series\'')))
  })
}
