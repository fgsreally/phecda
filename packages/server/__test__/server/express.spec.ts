import { describe, expect, it, vi } from 'vitest'
import request from 'supertest'
import express from 'express'
import type { ExpressCtx, Options } from '../../src/server/express'
import { bindApp } from '../../src/server/express'
import { ERROR_SYMBOL, Factory, addGuard, addInterceptor, addPipe, addPlugin } from '../../src'
import { Test } from '../fixtures/test.controller'

async function createServer(opts?: Options) {
  const data = await Factory([Test])
  const app = express()
  app.use(express.json())
  bindApp(app, data, opts)
  return app
}

describe('express ', () => {
  it('basic request', async () => {
    const app = await createServer()
    const res1 = await request(app).get('/get')
    expect(res1.body.msg).toBe('test')
    const res2 = await request(app).post('/post/phecda?id=1').send({ name: 'server' })
    expect(res2.text).toBe('phecda-server-1')

    const res3 = await request(app).post('/__PHECDA_SERVER__').send([
      {
        tag: 'Test-post',
        args: ['phecda', 'server', '1'],
      }, {
        tag: 'Test-get',
        args: [],
      },

    ])
    expect(res3.body[0]).toEqual('phecda-server-1')
    expect(res3.body[1]).toEqual({ msg: 'test' })
  })

  it('exception filter', async () => {
    const app = await createServer()

    const res1 = await request(app).get('/error')
    expect(res1.body).toEqual({ description: 'Http exception', message: 'test error', status: 500, [ERROR_SYMBOL]: true })
  })
  it('Pipe', async () => {
    const app = await createServer()

    addPipe('add', ({ arg }) => {
      return arg + 1
    })

    const res1 = await request(app).post('/pipe').send({ info: { name: '' } })

    expect(res1.body).toMatchObject({ message: 'name should be phecda', [ERROR_SYMBOL]: true })

    const res2 = await request(app).post('/pipe').query({ id: '1' }).send({ info: { name: 'phecda' } })
    expect(res2.text).toBe('11-phecda')
  })

  it('plugin', async () => {
    const fn = vi.fn()

    addPlugin('p1', (_req: Request, _res: Response, next: () => void) => {
      fn()
      next()
    })

    const app = await createServer({ plugins: ['p1'] })

    await request(app).get('/plugin')
    expect(fn).toHaveBeenCalledTimes(1)
    await request(app).post('/__PHECDA_SERVER__').send([{
      tag: 'Test-plugin',
      args: [],
    }])

    expect(fn).toHaveBeenCalledTimes(2)
  })

  it('guard/interceptor', async () => {
    const InterceptFn = vi.fn((str: string) => str)
    const Guardfn = vi.fn((str: string) => str)

    addGuard('g1', ({ params, index }: ExpressCtx) => {
      Guardfn('g1')

      if (index === undefined && params.test !== 'test')
        return false
      return true
    })
    addGuard('g2', () => {
      Guardfn('g2')
      return true
    })

    const mockInterceptor = () => {
      InterceptFn('start')
      return () => {
        InterceptFn('end')
      }
    }
    addInterceptor('i1', mockInterceptor)
    addInterceptor('i2', mockInterceptor)

    const app = await createServer({
      globalGuards: ['g2'],
      globalInterceptors: ['i2'],
    })
    const res1 = await request(app).post('/aop/no')
    expect(res1.body).toMatchObject({
      message: 'Guard exception--g1',
      status: 403,
      [ERROR_SYMBOL]: true,
    })
    await request(app).post('/aop')
    expect(Guardfn).toHaveBeenCalledTimes(2)

    await request(app).post('/__PHECDA_SERVER__').send(
      [
        {
          tag: 'Test-aop',
          args: ['test1'],
        },
        {
          tag: 'Test-aop',
          args: ['test2'],
        },
      ],

    )

    expect(InterceptFn).toHaveBeenCalledTimes(8)

    expect(Guardfn).toHaveBeenCalledTimes(6)
  })
  it('ctx', async () => {
    addGuard('g', (ctx: ExpressCtx) => {
      expect({ body: ctx.body, query: ctx.query, params: ctx.params }).toMatchSnapshot()
      return true
    })

    const app = await createServer({
      globalGuards: ['g'],
    })

    await request(app).post('/all/test?id=1').send({ name: 'test' }).expect(200, ['test', { name: 'test' }, '1'])
    await request(app).post('/__PHECDA_SERVER__').send(
      [
        {
          tag: 'Test-all',
          args: ['test', { name: 'test' }, '1'],
        },
        {
          tag: 'Test-all',
          args: ['test', { name: 'test' }, '2'],
        },
      ],

    ).expect(200)
  })
})
