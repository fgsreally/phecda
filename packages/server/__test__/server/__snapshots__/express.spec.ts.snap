// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`express  > ctx 1`] = `
{
  "body": {
    "name": "test",
  },
  "params": {
    "test": "test",
  },
  "query": {
    "id": "1",
  },
}
`;

exports[`express  > ctx 2`] = `
{
  "body": {
    "name": "test",
  },
  "params": {
    "test": "test",
  },
  "query": {
    "id": "1",
  },
}
`;

exports[`express  > ctx 3`] = `
{
  "body": {
    "name": "test",
  },
  "params": {
    "test": "test",
  },
  "query": {
    "id": "2",
  },
}
`;
